File: runme/mcgen/local/datan/datsrc/sdbinm.c
#include <math.h>
#include "datsrc.h"

double LIBFUNC sdbinm_(integer const P_T *k, integer const P_T *n,
                      double const P_T *p)
{
    /* System generated locals */
    double ret_val, d__1, d__2, d__3;

    /* Local variables */
    double ak, an, albinc, alfact, absarg, arg;

    ak = (double) (*k);
    an = (double) (*n);
    if (*n < 1 || *k < 0 || *k > *n) {
        ret_val = 0.;
    } else {
        d__1 = an + 1.;
        d__2 = ak + 1.;
        d__3 = an - ak + 1.;
        albinc = glngam_(&d__1) - glngam_(&d__2) - glngam_(&d__3);
        if (*p <= 0.) {
            if (*k == 0) {
                ret_val = 1.;
            } else {
                ret_val = 0.;
            }
        } else if (*p >= 1.) {
            if (*k == *n) {
                ret_val = 1.;
            } else {
                ret_val = 0.;
            }
        } else {
            alfact = log(*p) * ak + log(1. - *p) * (an - ak);
            arg = alfact + albinc;
            absarg = abs(arg);
            if (absarg < 1e-4) {
                ret_val = 1.;
            } else if (absarg > 100.) {
                ret_val = 0.;
            } else {
                ret_val = exp(arg);
            }
        }
    }
    return ret_val;
} /* sdbinm_ */

File: runme/mcgen/local/datan/datsrc/sdbinm.f
      DOUBLE PRECISION FUNCTION SDBINM(K,N,P)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      PARAMETER(ZERO=0.D0,ONE=1.D0,SMALL=.0001D0,BIG=100D0)
      AK=DBLE(K)
      AN=DBLE(N)
      IF(N.LT.1 .OR. K.LT.0 .OR. K.GT.N) THEN
        SDBINM=ZERO
      ELSE
        ALBINC=GLNGAM(AN+ONE)-GLNGAM(AK+ONE)-GLNGAM(AN-AK+ONE)
        IF(P.LE.ZERO) THEN
          IF(K.EQ.0) THEN
            SDBINM=ONE
          ELSE
            SDBINM=ZERO
          END IF
        ELSE IF(P.GE.ONE) THEN
          IF(K.EQ.N) THEN
            SDBINM=ONE
          ELSE
            SDBINM=ZERO
          END IF
        ELSE
          ALFACT=LOG(P)*AK+LOG(ONE-P)*(AN-AK)
          ARG=ALFACT+ALBINC
          ABSARG=ABS(ARG)
          IF(ABSARG.LT.SMALL) THEN
            SDBINM=ONE
          ELSE IF(ABSARG.GT.BIG) THEN
            SDBINM=ZERO
          ELSE
            SDBINM=EXP(ARG)
          END IF
        END IF
      END IF
      END
